x-pack/plugins/endpoint/public/embeddables/resolver/lib/tree_sequencers.ts

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// TODO: type root and children
+export function* depthFirstPreorder({ root, children }) {
```

Comment:
now that we have typescript, I stopped putting my arguments in objects like that. typescript can give you hints as to what each is, along w/ inline documentation. also it'll catch you if you're wrong in most situations. plus you can automatically refactor it across the whole project.

Reply:
as far as types go, something like this should work. I *think* the iterator type should be implied.
```ts
export function* depthFirstPreorder<T>(root: T, children: (parent: T) => T) {
```

Reply:
Here's a version that does away with the object wrappers in the arguments and has types:

```ts
export function* depthFirstPreorder<T>(root: T, children: (parent: T) => T[]): Iterable<T> {
  const nodesToVisit = [root];
  let currentNode
  while (currentNode = nodesToVisit.shift()) {
    nodesToVisit.unshift(...(children(currentNode) || []));
    yield currentNode;
  }
}

export function* levelOrder<T>(root: T, children: (parent: T) => T[]): Iterable<T> {
  let level = [root];
  while (level.length !== 0) {
    let nextLevel = [];
    for (const node of level) {
      yield node;
      nextLevel.push(...(children(node) || []));
    }
    level = nextLevel;
    nextLevel = [];
  }
}
```

Reply:
Thanks. Done.